### ISSUE TITLE
search service: allow prefix value searches (and add lots of tests) [phab:D153]

### ISSUE THREAD

**Issue by kythe** [google/kythe#360]
_Created by **schroederc** at 2015-03-31 21:00:13_: 

This change allows for some extra search possibilities like searching for all "variable" nodes with "local/**" subkinds or searching for a particular C++ "name" node without knowing its suffix type.

The `kythe` defaults the `suffix_wildcard` to be "%" for ease of use.

```
# all local variables
kythe search /kythe/node/kind variable /kythe/subkind local%

# files below a directory
kythe search --path kythe/java/com/google/devtools/kythe/analyzers/base/% /kythe/node/kind file

# search for the C++ Action name node (without knowing it is a class)
kythe search --lang c++ --sig Action:% /kythe/node/kind name
```

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-31 22:21:04_:

I'm curious about this interface. Why is this `bytes`? Components of a VName are all `string`. Does it make sense to say that `bytes` are the suffix of a `string`? Would it make less sense to add a flag to each `Fact` or `VName` component to request prefix matching?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-03-31 22:24:41_:

I thought about adding a flag per component but felt that it would complicate the API too much.

We could make this field a string, but then we'd go the other way for fact values which are bytestrings.  Maybe that would be better.

**Comment by kythe** [comment]
_**zarko** commented inline (context lost during migration) at 2015-03-31 22:30:50_:

For facts, it wouldn't really complicate the API so much. For VName components, you'd be adding five extra boolean fields, but I think that's actually easier to deal with as an API consumer.

If I'm a pedantic user, I would have to dynamically invent a wildcard for each search request, careful to make sure that it isn't inadvertently the suffix of any of the fields I don't want prefix-matching on. It seems much simpler to set the relevant flags.

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-03-31 23:10:43_:

Add prefix fields per Fact value and a VName mask for partial component prefixes

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2015-03-31 23:10:50_:

Done

**Comment by kythe** [comment]
_**zarko** wrote at 2015-04-01 00:28:13_:

Awesome, thanks!

**Comment by kythe** [comment]
_**zarko** accepted this revision at 2015-04-01 00:28:13_

**Comment by kythe** [comment]
_**schroederc** wrote at 2015-04-01 20:45:50_:

Rebase

**Comment by kythe** [comment]
_**schroederc** closed via commit b71ebd97ba557a4359a3706ca2c39a8397515a18 at 2015-04-01 21:07:10_
